### ISSUE TITLE
xmlSec: xmlSecDSigCtxDebugDump Access Violation

### ISSUE THREAD

**Issue by hafedh-trimeche** [microsoft/vcpkg#34932]
**Describe the bug**
xmlSecDSigCtxDebugDump Acces Violation.

**Environment**
- OS: Windows 11
- Compiler: vcpkg package management program version 2023-10-18-27de5b69dac4b6fe8259d283cd4011e6d20a84ce

**To Reproduce**
1. xmlSecDSigCtxSign(dsigCtx,SignatureNode);
2. xmlSecDSigCtxDebugDump 

**Expected behavior**
xmlSecDSigCtxDebugDump reports a signature log.

**Comment by JonLiu1993** [comment]
@hafedh-trimeche, Thanks for posting this issue, please provide your detailed error reproduction steps and error logs.

**Comment by hafedh-trimeche** [comment]
Hi,

Please find this log: **xmlsec 1.3.1#1**
```
:00007FFBB708090D ; C:\Windows\System32\ucrtbased.dll
:00007FFBB7036966 ; C:\Windows\System32\ucrtbased.dll
:00007FFBB704E3CA ; C:\Windows\System32\ucrtbased.dll
:00007FFBB704E4C8 ; C:\Windows\System32\ucrtbased.dll
:00007FFBB704F17A ; C:\Windows\System32\ucrtbased.dll
:00007FFBB707912C ; C:\Windows\System32\ucrtbased.dll
:00007FFBD9377CC0 ; D:\xmlsec\libxmlsec1.dll
:00007FFBD9377D30 ; D:\xmlsec\libxmlsec1.dll
:00007FFBD93C525E ; D:\xmlsec\libxmlsec1.dll
uXMLSec.DumpDSigDebug($1E6BCAD38D0)
```
and this one:
```
[00007FFBB7036966] Unknown function at ferror + $476
[00007FFBB704E3CA] Unknown function at __p__commode + $DDA
[00007FFBB704E4C8] Unknown function at __p__commode + $ED8
[00007FFBB704F17A] Unknown function at __p__commode + $1B8A
[00007FFBB707912C] __stdio_common_vfprintf + $5C
[00007FFBDB107CC0] Unknown function at xmlSecSimpleKeysStoreAdoptKey + $25EB5
[00007FFBDB107D30] Unknown function at xmlSecSimpleKeysStoreAdoptKey + $25F25
[00007FFBDB15525E] Unknown function at xmlSecSimpleKeysStoreAdoptKey + $73453
[0000000001780B26] uXMLSec.DumpDSigDebug (Line 175, "uXMLSec.pas" + 2) + $14
```
uXMLSec.DumpDSigDebug is calling:
`xmlSecDSigCtxDebugDump(dsigCtx,xmlSecDSigOutput)
`

**Comment by JonLiu1993** [comment]
@hafedh-trimeche, My computer is 64 bit and I can build xmlSecport successfully:
````
printf("sizeof time_t is: %d\n", sizeof(time_t));
````
output:
````
sizeof time_t is: 8
````

**Comment by JonLiu1993** [comment]
@hafedh-trimeche, Does this problem still exist?

**Comment by hafedh-trimeche** [comment]
Yes it does: xmlsec 1.3.1#2

**Comment by JonLiu1993** [comment]
xmlSec has the latest version 1.3.2. I need to update it. Can you try it?

**Comment by hafedh-trimeche** [comment]
Sorry I can't!
I'm using **VCPKG** which automatically generates library.
Waiting for the next version.

**Comment by lsh123** [comment]
I don't believe changes in 1.3.2 would resolve this problem. Overall, this looks like one of the two possible issues for me:
- mismatch in the C runtime for various libraries (xmlsec, libxml2, libxslt, openssl, ... )
- mismatch between compiled and loaded libraries (i.e. library versions, compilation flags, etc are not the same)

**Comment by JonLiu1993** [comment]
@hafedh-trimeche, The latest xmlsec 1.3.4 can solved your issue?

**Comment by hafedh-trimeche** [comment]
Hello,
Even VCPKG updated, xmlsec still at version 1.3.3:
![image](https://github.com/microsoft/vcpkg/assets/988584/52587456-cf5d-46b6-958d-57a20fe3339a)

With the same result!

**Comment by dg0yt** [comment]
Did you ever check if you call the function with the right argument types, build with the right link lib, run with the right runtime lib?
IIUC the call is coming from pascal (`Line 175, "uXMLSec.pas" + 2`). Not sure if there aren't more traps at the language boundary.

**Comment by hafedh-trimeche** [comment]
**xmlSecDSigCtxDebugDump** has only two (2) parameters:
**dsigCtx** & **output** which are pointers.
**output** is open using **fopen('xmlSec.log','w')** and shouldn't cause problem.
It seams that the library doesn't check some **null values** before dumping related content. 
The calling convention doesn't matter because the library is a 64 bit one.
Best regards.

**Comment by hafedh-trimeche** [comment]
I used **fopen** and **fclose** function statically linked to open and close Log file:
https://github.com/tinyBigGAMES/CPas/blob/main/lib/CPas.CRuntime.pas

This exception is raised:
```
OpenSSL: FATAL
---------------------------
OPENSSL_Uplink(00007FFD4F3DCA38,08): no OPENSSL_Applink
```

Is there a way to log content to a memory stream (buffer) not to a file?

**Comment by lsh123** [comment]
If you can reproduce the issue with xmlsec command line utility then it would be easier to debug it.  Use ```--print-debug``` flag to trigger xmlSecDSigCtxDebugDump usage. As of now, I stand by my comment above about mismatch between headers / compile flags and library version. Using xmlsec command line utility would guarantee that compilation / linking is consistent.

**Comment by lsh123** [comment]
oh I missed it: "I used fopen and fclose function statically linked to open and close Log file". Of course this is going to crash. Welcome to Microsoft runtimes hell. Question 2.7 in the FAQ (https://www.aleksey.com/xmlsec/faq.html).

**Comment by dg0yt** [comment]
> oh I missed it: "I used fopen and fclose function statically linked to open and close Log file". Of course this is going to crash. Welcome to Microsoft runtimes hell. Question 2.7 in the FAQ (https://www.aleksey.com/xmlsec/faq.html).

Well, runtimes hell is what vcpkg tries to eliminate AFAIU. There are even post-build checks. So I wouldn't be surprised if executables from the ports just work.  
But does the user project use the same configuration as the ports? 
And does OPENSSL_Applink have a meaning here?
